src/Compiler/Microsoft.CodeAnalysis.Razor/src/DefaultTypeNameFeature.cs

```diff
@@ -22,24 +22,45 @@ public override IReadOnlyList<string> ParseTypeParameters(string typeName)
         }
 
         var parsed = SyntaxFactory.ParseTypeName(typeName);
+
         if (parsed is IdentifierNameSyntax identifier)
         {
             return Array.Empty<string>();
         }
-        else if (parsed is ArrayTypeSyntax array)
+
+        if (TryParseCore(parsed) is { } list)
         {
-            return new[] { array.ElementType.ToString() };
+            return list;
         }
-        else if (parsed is TupleTypeSyntax tuple)
+
+        return parsed.DescendantNodesAndSelf()
+            .OfType<TypeArgumentListSyntax>()
+            .SelectMany(arg => arg.Arguments)
+            .SelectMany(ParseCore).ToArray();
+
+        static IReadOnlyList<string> TryParseCore(TypeSyntax parsed)
         {
-            return tuple.Elements.Select(a => a.ToString()).ToList();
+            if (parsed is ArrayTypeSyntax array)
+            {
+                return new[] { array.ElementType.ToString() };
```

Comment:
Are we going to have an issue with arrays of arrays, or arrays of tuples?

Reply:
Cannot add an integration test to verify that, because:
- If we get `T[][]`, the `array.ElementType` is still `T`.
- If we get `T[]?[]`, this method gets it stripped as `T[][]`.
- Type inference for components doesn't seem to work with tuples.

But otherwise you're right. Fixed and added a unit test.